sql: fix portals after exhausting rows

yuzefovich opened the pull request:
Previously, we would erroneously restart the execution from the very
beginning of empty, unclosed portals after they have been fully
exhausted when we should be returning no rows or an error in such
scenarios. This is now fixed by tracking whether a portal is exhausted
or not and intercepting the calls to `execStmt` when the conn executor
state machine is in an open state.

Note that the current solution has known deviations from Postgres:
- when attempting to execute portals of statements that don't return row
sets, on the second and consequent attempt PG returns an error while we
are silently doing nothing (meaning we do not run the statement at all
and return 0 rows)
- we incorrectly populate "command tag" field of pgwire messages of some
rows-returning statements after the portal suspension (for example,
a suspended UPDATE RETURNING in PG will return the total count of "rows
affected" while we will return the count since the last suspension).

These deviations are deemed acceptable since this commit fixes a much
worse problem - re-executing an exhausted portal (which could be
a mutation meaning, previously, we could have executed a mutation
multiple times).

The reasons for why this commit does not address these deviations are:
- Postgres has a concept of "portal strategy"
(see https://github.com/postgres/postgres/blob/2f9661311b83dc481fc19f6e3bda015392010a40/src/include/utils/portal.h#L89).
- Postgres has a concept of "command" type (these are things like
SELECTs, UPDATEs, INSERTs, etc,
see https://github.com/postgres/postgres/blob/1aac32df89eb19949050f6f27c268122833ad036/src/include/nodes/nodes.h#L672).

CRDB doesn't have these concepts, and without them any attempt to
simulate Postgres results in a very error-prone and brittle code.

Fixes: #48448.

Release note (bug fix): Previously, CockroachDB would erroneously
restart the execution of empty, unclosed portals after they have been
fully exhausted, and this has been fixed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48842)
<!-- Reviewable:end -->

yuzefovich commented:
cc @jordanlewis @asubiotto

jordanlewis commented on pkg/sql/conn_executor_exec.go:
```diff
@@ -440,10 +440,10 @@ func (ex *connExecutor) execStmtInOpenState(
 	// change would forget to add the call).
 	//
-	// TODO(andrei): really the code should be re-architectued to ensure
+	// TODO(andrei): really the code should be re-architectured to ensure
```
nit: re-architected* :)

jordanlewis commented on pkg/sql/conn_executor.go:
```diff
@@ -1406,7 +1406,29 @@ func (ex *connExecutor) execCmd(ctx context.Context) error {
+		case stateOpen:
+			// We're about the execute the statement in an open state which
+			// could trigger the dispatch to the execution engine. However, it
+			// is possible that we're trying to execute already exhausted
```
nit: execute an already exhausted*

jordanlewis commented on pkg/sql/conn_executor.go:
```diff
@@ -1406,7 +1406,29 @@ func (ex *connExecutor) execCmd(ctx context.Context) error {
+			// We're about the execute the statement in an open state which
+			// could trigger the dispatch to the execution engine. However, it
+			// is possible that we're trying to execute already exhausted
+			// portal - in such scenario we should return no rows, but the
```
nit: in such a scenario*

jordanlewis commented on pkg/sql/conn_executor.go:
```diff
@@ -1406,7 +1406,29 @@ func (ex *connExecutor) execCmd(ctx context.Context) error {
-		ev, payload, err = ex.execStmt(stmtCtx, curStmt, stmtRes, pinfo)
+		switch ex.machine.CurState().(type) {
+		case stateOpen:
+			// We're about the execute the statement in an open state which
```
nit: we're about to execute

jordanlewis commented on pkg/sql/conn_executor.go:
```diff
@@ -1406,7 +1406,29 @@ func (ex *connExecutor) execCmd(ctx context.Context) error {
+			// nothing to return on an exhausted portal.
+			if !portal.exhausted {
+				ev, payload, err = ex.execStmt(stmtCtx, curStmt, stmtRes, pinfo)
+				// The query execution has been complete (because we only
```
nit: has completed

jordanlewis reviewed (approved):
Looks good, I think. Are there no other places where we could execute and exhaust a portal?

I'm concerned that the failures in teamcity could be related, though they look quite unrelated. Can you reproduce them locally?

yuzefovich commented:
I can reproduce failures locally, so it must be this change to blame, but I'm at a loss at what's wrong, I was hoping you guys would see the reason :)

asubiotto reviewed (commented):
Nice! I have no insight re the logic test failures

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-:-M7IUMNz0kt_9KwI2kWM:begq9a8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @mjibson, and @yuzefovich)

___
*[pkg/sql/conn_executor.go, line 1424 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7IUIOFDE4tkqEkWD-4:-M7IUIOFDE4tkqEkWD-5:bqq72ka) ([raw file](https://github.com/cockroachdb/cockroach/blob/3524bd915ba2f181e21749153fe1a9866dbe3084/pkg/sql/conn_executor.go#L1424)):*
> ```Go
> 				ev, payload, err = ex.execStmt(stmtCtx, curStmt, stmtRes, pinfo)
> 				// The query execution has been complete (because we only
> 				// supports execution of portals to completion) which means
> ```

nit: `s/supports/support`


<!-- Sent from Reviewable.io -->

andreimatei reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48842)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @mjibson, and @yuzefovich)

___
*[pkg/sql/conn_executor.go, line 1409 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7IaZwQ-9l4pxM7HBCT:-M7IaZwQ-9l4pxM7HBCU:bs0ugzu) ([raw file](https://github.com/cockroachdb/cockroach/blob/3524bd915ba2f181e21749153fe1a9866dbe3084/pkg/sql/conn_executor.go#L1409)):*
> ```Go
> 		}
> 		stmtCtx := withStatement(ctx, ex.curStmt)
> 		switch ex.machine.CurState().(type) {
> ```

Have you putting the setting off the exhausted flag inside `limitedCommandResult`? That's the guy that know whether a portal is exhausted or not.

The checking of whether a portal is exhausted I'd try to put in `execStmtInOpenState()` so you don't have to inspect the state at this level. Maybe add an `Exhausted()` method on the `RestrictedCommandResult()` interface.
___
*[pkg/sql/conn_executor.go, line 1427 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7IeZufC-uOUDuQQj1v:-M7IeZufC-uOUDuQQj1w:b-m73500) ([raw file](https://github.com/cockroachdb/cockroach/blob/3524bd915ba2f181e21749153fe1a9866dbe3084/pkg/sql/conn_executor.go#L1427)):*
> ```Go
> 				// that all rows from this portal have been returned - the
> 				// portal is now exhausted.
> 				portal.exhausted = true
> ```

Do we need some logic to reset this field when doing auto-retries? See `prepStmtNamespace.resetTo()`. Please add a test showing that an auto-retry works in re-executing a portal.
I think before this patch portals were immutable so it was easy to share instances between the set of portals captured at the "rewind position" and the current set of portals. But if they're not immutable, more care I think is needed. 


<!-- Sent from Reviewable.io -->

mjibson commented:
The failures are from some jobs tests: `jobs_test.go:1836: sql: no rows in result set` after running `CANCEL JOB`, which has a `RowsAffected` return.

The current test only tests a `Rows` StatementType. Add tests that exercise RowsAffected, Ack, and DDL to ensure we're handling this stuff correctly. You can use the `-rewrite -addr localhost:5432` flags in the pgtest stuff to test against a real postgres server and generate expected results. I suspect doing this will illustrate the problem more clearly.

mjibson commented on pkg/sql/pgwire/testdata/pgtest/portals:
```diff
@@ -342,3 +342,151 @@ ReadyForQuery
+{"Type":"BindComplete"}
+{"Type":"CommandComplete","CommandTag":"ALTER ROLE"}
+{"Type":"ReadyForQuery","TxStatus":"T"}
+
```
Add a statement type that returns a ROWS type, probably an UPDATE. Execute that twice and then SELECT from it to verify that the UPDATE was only happened once. So it'd probably have to be like `UPDATE t SET s = s + 1` or something that's not idempotent.

yuzefovich reviewed (commented):
I fixed up the first commit, and now the CI is green. The issue was not resetting portal's newly added `exhausted` flag during auto-retries. The PR is missing a test for auto-retries with re-executing the portal, so it is still in a WIP mode, but ready for another look.

I think fundamentally the issue here is that the main state machine is not aware of existence of a partial "side" state machine implemented by `limitedCommandResult` which was a known limitation during the portals' implementation. We probably have two choices here:
1. do a hacky solution that I'm proposing here by adding some additional tracking info and teaching the conn executor's state machine about an "exhausted" state of a portal - this is done in the first commit. The benefit is that it's a relatively minor change, but we're making the whole situation with portals handling slightly worse.
2. perform a fundamental rewrite of how portals are handled. This will probably require rewriting the conn executor's state machine so that it supported what `limitedCommandResult` gives us. I'm not sure how big of a lift that would be, but I think it was considered during initial portals implementation and probably deemed too big of an investment.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-:-MBf0Bwz5ZI2nE5MhmEs:b-4ryq03)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @mjibson)

___
*[pkg/sql/conn_executor.go, line 1409 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7IaZwQ-9l4pxM7HBCT:-MBeqnFZ1qIIE0HRUv-P:b-iq9pco) ([raw file](https://github.com/cockroachdb/cockroach/blob/3524bd915ba2f181e21749153fe1a9866dbe3084/pkg/sql/conn_executor.go#L1409)):*
> Have you putting the setting off the exhausted flag inside `limitedCommandResult`? That's the guy that know whether a portal is exhausted or not.

I have thought about it, but I don't think it is possible update the `exhausted` flag within the command result because that thing doesn't know when there will be no more rows to add using `AddRow`. I think it is known only once `dispatchToExecutionEngine` method returns.

We could add another method to `RestrictedCommandResult` interface, something like `SetExhausted` which would notify the command result that the execution engine has returned all of the rows, that would occur in `execWithDistSQLEngine`, after `PlanAndRun` call if there is no error. What do you think about that?

> The checking of whether a portal is exhausted I'd try to put in `execStmtInOpenState()` so you don't have to inspect the state at this level. Maybe add an `Exhausted()` method on the `RestrictedCommandResult()` interface.

Hm, I prototyped both things in the second commit, but now the regression test fails :/ Maybe you could see what I'm doing wrong?
___
*[pkg/sql/conn_executor.go, line 1427 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7IeZufC-uOUDuQQj1v:-MBf-K2w71fZias269FA:b-arlrr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3524bd915ba2f181e21749153fe1a9866dbe3084/pkg/sql/conn_executor.go#L1427)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Do we need some logic to reset this field when doing auto-retries? See `prepStmtNamespace.resetTo()`. Please add a test showing that an auto-retry works in re-executing a portal.
I think before this patch portals were immutable so it was easy to share instances between the set of portals captured at the "rewind position" and the current set of portals. But if they're not immutable, more care I think is needed. 
</blockquote></details>

Yes, you were right - missing logic for auto-retries handling was the cause of failures we saw on CI earlier. Could you give me a pointer to how I would write such test? I see `conn_executor_internal_test.go` which mocks out a `connExecutor`, but I'm not sure how I would trigger an auto-retry.
___
*[pkg/sql/pgwire/testdata/pgtest/portals, line 482 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7FRxi84OqPVll_sUMG-r2-482:-MBf-pRG1BXTapd1CLOS:b-r5ae36) ([raw file](https://github.com/cockroachdb/cockroach/blob/da502756fb5180b74598c11de499fe06eff58656/pkg/sql/pgwire/testdata/pgtest/portals#L482)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Add a statement type that returns a ROWS type, probably an UPDATE. Execute that twice and then SELECT from it to verify that the UPDATE was only happened once. So it'd probably have to be like `UPDATE t SET s = s + 1` or something that's not idempotent.
</blockquote></details>

Thanks, added.

However, there is one thing I can't figure out (in the first commit) that maybe you could spot a problem with. Namely, if the statement is of "RowsAffected" type (an update without RETURNING), then I'm hitting an error when executing the portal for the second time running against postgres.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48842)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @mjibson)

___
*[pkg/sql/pgwire/testdata/pgtest/portals, line 482 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7FRxi84OqPVll_sUMG-r2-482:-MBk-0OQ8I1S-pEIKpAy:b-vtn7sf) ([raw file](https://github.com/cockroachdb/cockroach/blob/da502756fb5180b74598c11de499fe06eff58656/pkg/sql/pgwire/testdata/pgtest/portals#L482)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Thanks, added.

However, there is one thing I can't figure out (in the first commit) that maybe you could spot a problem with. Namely, if the statement is of "RowsAffected" type (an update without RETURNING), then I'm hitting an error when executing the portal for the second time running against postgres.
</blockquote></details>

Ok, so we've updated the test so that an UPDATE without RETURNING in a portal can only be executed once and returns an error on the second try, but an UPDATE with RETURNING can be attempted to be executed multiple times yet only the first one actually runs.

However, in the former case the "fix" in the first commit exhibits a different behavior:
```
        datadriven.go:118: testdata/pgtest/portals:523: waiting for *pgproto3.ErrorResponse, got &pgproto3.ReadyForQuery{TxStatus:0x54}
```
meaning that we don't return an error, and, instead, do nothing:
```
until
ReadyForQuery
----
{"Type":"CommandComplete","CommandTag":"UPDATE 0"}
{"Type":"ReadyForQuery","TxStatus":"T"}
```
The UPDATE is actually only executed once. It seems like if we want to also return an error in such scenario, more thought is needed.


<!-- Sent from Reviewable.io -->

mjibson commented on pkg/sql/pgwire/testdata/pgtest/portals:
```diff
@@ -342,3 +342,268 @@ ReadyForQuery
+
+send
+Execute {"MaxRows": 2}
+Sync
```
Remove all the `Sync` after `Execute`, this will remove some of the double `ReadyForQuery` things you have around.

mjibson commented on pkg/sql/pgwire/testdata/pgtest/portals:
```diff
@@ -342,3 +342,268 @@ ReadyForQuery
+{"Type":"CommandComplete","CommandTag":"DROP ROLE"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+# Execute a statement of "RowsAffected" statement type. We will try to execute
```
Add a test that's a copy of this but does an ALTER/CREATE/DROP or something to assert we do whatever postgres does in that case too.

mjibson commented on pkg/sql/prepared_stmt.go:
```diff
@@ -121,6 +121,16 @@ type preparedStatementsAccessor interface {
 // References need to be registered with incRef() and de-registered with
 // decRef().
 type PreparedPortal struct {
+	*immutablePreparedPortal
```
What's the reason we need to separate out the types here? Why can't we add exhausted to the existing PreparedPortal type?

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48842)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @mjibson)

___
*[pkg/sql/conn_executor.go, line 1409 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7IaZwQ-9l4pxM7HBCT:-MBojD4Z36YFB4GIf3ok:b3nyemp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3524bd915ba2f181e21749153fe1a9866dbe3084/pkg/sql/conn_executor.go#L1409)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

> Have you putting the setting off the exhausted flag inside `limitedCommandResult`? That's the guy that know whether a portal is exhausted or not.

I have thought about it, but I don't think it is possible update the `exhausted` flag within the command result because that thing doesn't know when there will be no more rows to add using `AddRow`. I think it is known only once `dispatchToExecutionEngine` method returns.

We could add another method to `RestrictedCommandResult` interface, something like `SetExhausted` which would notify the command result that the execution engine has returned all of the rows, that would occur in `execWithDistSQLEngine`, after `PlanAndRun` call if there is no error. What do you think about that?

> The checking of whether a portal is exhausted I'd try to put in `execStmtInOpenState()` so you don't have to inspect the state at this level. Maybe add an `Exhausted()` method on the `RestrictedCommandResult()` interface.

Hm, I prototyped both things in the second commit, but now the regression test fails :/ Maybe you could see what I'm doing wrong?
</blockquote></details>

I moved that second commit to a different branch, it is here https://github.com/yuzefovich/cockroach/commit/1c7b54082de3aa6afb6219a4db85dc8616d174b0.
___
*[pkg/sql/prepared_stmt.go, line 124 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7FRxi84OqPVll_sUMH-r3-124:-MBohZwH4oOuMz6MZ17N:b-zbw0et) ([raw file](https://github.com/cockroachdb/cockroach/blob/da8fc14dc971e1cfd142a5f377d653ca7421c2fc/pkg/sql/prepared_stmt.go#L124)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

What's the reason we need to separate out the types here? Why can't we add exhausted to the existing PreparedPortal type?
</blockquote></details>

Added a comment. Basically I think we want to have multiple `PreparedPortal`s share the same "immutable core" struct, but we also want to have separate `exhausted` variable so that we could rewind the portal correctly.
___
*[pkg/sql/pgwire/testdata/pgtest/portals, line 376 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7FRxi84OqPVll_sUMG-r3-376:-MBoWlrD96EvzOAyConi:b-2d7rkp) ([raw file](https://github.com/cockroachdb/cockroach/blob/da8fc14dc971e1cfd142a5f377d653ca7421c2fc/pkg/sql/pgwire/testdata/pgtest/portals#L376)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Remove all the `Sync` after `Execute`, this will remove some of the double `ReadyForQuery` things you have around. 
</blockquote></details>

I don't think we can remove `Sync`s. Here is a quote from Postgres docs:
`At completion of each series of extended-query messages, the frontend should issue a Sync message. This parameterless message causes the backend to close the current transaction if it's not inside a BEGIN/COMMIT transaction block ("close" meaning to commit if no error, or roll back if error). Then a ReadyForQuery response is issued.`
Also, there are `Sync`s after `Execute`s in this file already.
___
*[pkg/sql/pgwire/testdata/pgtest/portals, line 489 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7FRxi84OqPVll_sUMG-r3-489:-MBo_Ap51bePYYMcxLGN:bn9phyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/da8fc14dc971e1cfd142a5f377d653ca7421c2fc/pkg/sql/pgwire/testdata/pgtest/portals#L489)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Add a test that's a copy of this but does an ALTER/CREATE/DROP or something to assert we do whatever postgres does in that case too.
</blockquote></details>

Done.

However, similar to an UPDATE without RETURNING, Postgres returns an error whereas we now silently do nothing. Still I think that the first commit in this PR is beneficial since it fixes a worse problem (that a portal can be executed multiple times mistakenly) at the cost of deviation in the behavior of handling of portals on statements that don't return row sets (such portals can only be executed to completion, and it appears that PG returns an error on consequent attempts whereas we'll be silently doing nothing). I moved such cases into `portals_crbugs`.
___
*[pkg/sql/pgwire/testdata/pgtest/portals, line 464 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MBoix5oE7MSkm0By0ee:-MBoix5oE7MSkm0By0ef:b-902p3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1adb0dd709b77e410997d597a3a993a5cab14/pkg/sql/pgwire/testdata/pgtest/portals#L464)):*
> ```
> ----
> {"Type":"DataRow","Values":[{"text":"30"}]}
> {"Type":"CommandComplete","CommandTag":"UPDATE 3"}
> ```

There is a difference between us and Postgres on the command tag here:
```
            testdata/pgtest/portals:460: ReadyForQuery
            expected:
            {"Type":"DataRow","Values":[{"text":"30"}]}
            {"Type":"CommandComplete","CommandTag":"UPDATE 3"}
            {"Type":"ReadyForQuery","TxStatus":"T"}
            
            found:
            {"Type":"DataRow","Values":[{"text":"30"}]}
            {"Type":"CommandComplete","CommandTag":"UPDATE 1"}
            {"Type":"ReadyForQuery","TxStatus":"T"}
```
Is this concerning?


<!-- Sent from Reviewable.io -->

madelynnblue reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48842)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @mjibson, and @yuzefovich)

___
*[pkg/sql/pgwire/testdata/pgtest/portals, line 376 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7FRxi84OqPVll_sUMG-r3-376:-MBolKnF513xcalZcFEO:bpn7vrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/da8fc14dc971e1cfd142a5f377d653ca7421c2fc/pkg/sql/pgwire/testdata/pgtest/portals#L376)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I don't think we can remove `Sync`s. Here is a quote from Postgres docs:
`At completion of each series of extended-query messages, the frontend should issue a Sync message. This parameterless message causes the backend to close the current transaction if it's not inside a BEGIN/COMMIT transaction block ("close" meaning to commit if no error, or roll back if error). Then a ReadyForQuery response is issued.`
Also, there are `Sync`s after `Execute`s in this file already.
</blockquote></details>

I meant remove `Sync` after `Query`, not `Execute`.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48842)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @mjibson)

___
*[pkg/sql/pgwire/testdata/pgtest/portals, line 376 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7FRxi84OqPVll_sUMG-r3-376:-MBolw9x6v7BE0c9iZDg:bucjicg) ([raw file](https://github.com/cockroachdb/cockroach/blob/da8fc14dc971e1cfd142a5f377d653ca7421c2fc/pkg/sql/pgwire/testdata/pgtest/portals#L376)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I meant remove `Sync` after `Query`, not `Execute`.
</blockquote></details>

Oh, that makes sense. I think I removed all such `Sync`s.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48842)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @mjibson, and @yuzefovich)

___
*[pkg/sql/conn_executor.go, line 1427 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7IeZufC-uOUDuQQj1v:-MC7e6_LA4hDdV7_ily1:b-1kvxx3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3524bd915ba2f181e21749153fe1a9866dbe3084/pkg/sql/conn_executor.go#L1427)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Yes, you were right - missing logic for auto-retries handling was the cause of failures we saw on CI earlier. Could you give me a pointer to how I would write such test? I see `conn_executor_internal_test.go` which mocks out a `connExecutor`, but I'm not sure how I would trigger an auto-retry.
</blockquote></details>

But now if we're resetting the `exhausted` field when we rewind, in `portalNamespace.resetTo()`, I think you can set it here unconditionally, no? And I think that's the correct thing to do - you don't want to allow this portal to be executed again after an error. Which I think it couldn't because, after an error, you have to either start a new txn or rollback to a savepoint (both of which muck with the portals), but still I think you want protection at this layer too.
___
*[pkg/sql/conn_executor.go, line 1443 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MC7gwxE3JS7dmn7VQFS:-MC7gwxF0XDw2SWjtcY3:bggu255) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1adb0dd709b77e410997d597a3a993a5cab14/pkg/sql/conn_executor.go#L1443)):*
> ```Go
> 			// to prevent such behavior, we check whether the portal has been
> 			// exhausted and execute the statement only if it hasn't. If it has
> 			// been exhausted, then we simply do nothing because there is
> ```

Is this how Postgres behaves - there's no message sent back when trying to execute an exhausted portal? No `CommandComplete` message?
___
*[pkg/sql/prepared_stmt.go, line 124 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7FRxi84OqPVll_sUMH-r3-124:-MC7YhVV0mXyfikkhV-8:bfv7568) ([raw file](https://github.com/cockroachdb/cockroach/blob/da8fc14dc971e1cfd142a5f377d653ca7421c2fc/pkg/sql/prepared_stmt.go#L124)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Added a comment. Basically I think we want to have multiple `PreparedPortal`s share the same "immutable core" struct, but we also want to have separate `exhausted` variable so that we could rewind the portal correctly.
</blockquote></details>

The current structure is confusing because you've left `incRef` as a method on `PreparedPortal` but it refers just to the `immutablePreparedPortal`. In `prepStmtNamespace.resetTo()` we now increment on a `p` and then proceed to make a copy of that `p`.
Since we're adding this extra state, I would try to get rid of any sharing in `prepStmtNamespace`; it's becoming too complicated to try to share some, but not all of the portal. So, instead of `map[string]*PreparedPortal`, I'd have `map[string]PreparedPortal`. You can't copy the `memAcc` but I think that having an account per portal is pretty silly and inefficient anyway. I'd put `prepStmtNamespace` in charge of accounting for all the portal - let it add and delete bytes from a single account as it registers/unregister portals.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48842)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @mjibson)

___
*[pkg/sql/conn_executor.go, line 1427 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7IeZufC-uOUDuQQj1v:-MCiOyjP3GkxHcdJaC_N:b-ks66n0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3524bd915ba2f181e21749153fe1a9866dbe3084/pkg/sql/conn_executor.go#L1427)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

But now if we're resetting the `exhausted` field when we rewind, in `portalNamespace.resetTo()`, I think you can set it here unconditionally, no? And I think that's the correct thing to do - you don't want to allow this portal to be executed again after an error. Which I think it couldn't because, after an error, you have to either start a new txn or rollback to a savepoint (both of which muck with the portals), but still I think you want protection at this layer too.
</blockquote></details>

I think you're right, done. This condition came from me not really understanding how `connExecutor` works - I can't say that I fully understand it, but I have a better grasp of it now.
___
*[pkg/sql/conn_executor.go, line 1443 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MC7gwxE3JS7dmn7VQFS:-MCiPCSm36Azv3GoSsyD:bvb59nr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1adb0dd709b77e410997d597a3a993a5cab14/pkg/sql/conn_executor.go#L1443)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Is this how Postgres behaves - there's no message sent back when trying to execute an exhausted portal? No `CommandComplete` message?
</blockquote></details>

Postgres does send `CommandComplete` with `SELECT 0` command tag (and we have such newly-added test case in `pgtest/portals:386-396`). Is the comment confusing? By "we do nothing" I meant that we need to not dispatch the query to the execution engine, but the remaining state transitions of `connExecutor` will still occur, only as if the execution engine returned 0 rows. Updated the comment.

What's different is that we are not returning an error in cases when Postgres would on attempts on executing exhausted portals for certain statement types. See my comment on `testdata/portals` file.
___
*[pkg/sql/prepared_stmt.go, line 124 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7FRxi84OqPVll_sUMH-r3-124:-MCiJTpRAVwzp52EeYnY:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/da8fc14dc971e1cfd142a5f377d653ca7421c2fc/pkg/sql/prepared_stmt.go#L124)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

The current structure is confusing because you've left `incRef` as a method on `PreparedPortal` but it refers just to the `immutablePreparedPortal`. In `prepStmtNamespace.resetTo()` we now increment on a `p` and then proceed to make a copy of that `p`.
Since we're adding this extra state, I would try to get rid of any sharing in `prepStmtNamespace`; it's becoming too complicated to try to share some, but not all of the portal. So, instead of `map[string]*PreparedPortal`, I'd have `map[string]PreparedPortal`. You can't copy the `memAcc` but I think that having an account per portal is pretty silly and inefficient anyway. I'd put `prepStmtNamespace` in charge of accounting for all the portal - let it add and delete bytes from a single account as it registers/unregister portals.
</blockquote></details>

Good point, done.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48842)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @mjibson, and @yuzefovich)

___
*[pkg/sql/conn_executor.go, line 1443 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MC7gwxE3JS7dmn7VQFS:-MCwZx5PBKT_lutgWRFz:b-brkqo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1adb0dd709b77e410997d597a3a993a5cab14/pkg/sql/conn_executor.go#L1443)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Postgres does send `CommandComplete` with `SELECT 0` command tag (and we have such newly-added test case in `pgtest/portals:386-396`). Is the comment confusing? By "we do nothing" I meant that we need to not dispatch the query to the execution engine, but the remaining state transitions of `connExecutor` will still occur, only as if the execution engine returned 0 rows. Updated the comment.

What's different is that we are not returning an error in cases when Postgres would on attempts on executing exhausted portals for certain statement types. See my comment on `testdata/portals` file.
</blockquote></details>

comment is better now

You're sure that postgres allows you to `Execute` a portal multiple times after the portal is suspended, right?
___
*[pkg/sql/conn_executor_prepare.go, line 401 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MCwY5y69r1YLhENzz9L:-MCwY5y69r1YLhENzz9M:bssfbgq) ([raw file](https://github.com/cockroachdb/cockroach/blob/44563956f320d4e86424b6f37919a4545f62b24f/pkg/sql/conn_executor_prepare.go#L401)):*
> ```Go
> 	if portal, ok := ex.extraTxnState.prepStmtsNamespace.portals[portalName]; !ok {
> 		panic(fmt.Sprintf("portal %s doesn't exist", portalName))
> 	} else {
> ```

nit: drop the `else` and unindent the expected case
___
*[pkg/sql/pgwire/testdata/pgtest/portals, line 346 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MCw_zI6Bh7PSceyRise:-MCw_zI6Bh7PSceyRisf:b-rt19jc) ([raw file](https://github.com/cockroachdb/cockroach/blob/44563956f320d4e86424b6f37919a4545f62b24f/pkg/sql/pgwire/testdata/pgtest/portals#L346)):*
> ```
> {"Type":"ReadyForQuery","TxStatus":"I"}
> 
> # Regression for #48448
> ```

instead of click-bait, please put some words about what we're testing
___
*[pkg/sql/pgwire/testdata/pgtest/portals, line 396 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MCwaBWm2vQSDkobFv94:-MCwaBWm2vQSDkobFv95:bfkzpl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/44563956f320d4e86424b6f37919a4545f62b24f/pkg/sql/pgwire/testdata/pgtest/portals#L396)):*
> ```
> {"Type":"CommandComplete","CommandTag":"SELECT 0"}
> {"Type":"ReadyForQuery","TxStatus":"T"}
> ```

I'd try another `Execute` here. I wanna make sure that pg returns `SELECT 0` multiple times.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48842)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @mjibson, and @yuzefovich)

___
*[pkg/sql/pgwire/testdata/pgtest/portals, line 464 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MBoix5oE7MSkm0By0ee:-MCwcQVs9oHSAELg4qGa:b-vv8yll) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1adb0dd709b77e410997d597a3a993a5cab14/pkg/sql/pgwire/testdata/pgtest/portals#L464)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

There is a difference between us and Postgres on the command tag here:
```
            testdata/pgtest/portals:460: ReadyForQuery
            expected:
            {"Type":"DataRow","Values":[{"text":"30"}]}
            {"Type":"CommandComplete","CommandTag":"UPDATE 3"}
            {"Type":"ReadyForQuery","TxStatus":"T"}
            
            found:
            {"Type":"DataRow","Values":[{"text":"30"}]}
            {"Type":"CommandComplete","CommandTag":"UPDATE 1"}
            {"Type":"ReadyForQuery","TxStatus":"T"}
```
Is this concerning?
</blockquote></details>

If this question is for me, I think this is the time to figure out the exact details of the pg implementation. Obviously, the protocol is only defined by the Postgres code :0
I don't understand what the diff here says, but I think the one below seems bad


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Alright, I think this PR is now in a pretty good shape - it fixes the bug and doesn't introduce any deviations from the postgres (at the cost of some hacky code). RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-:-MDHh77G-iMa4VffDh33:bkxe87y)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @mjibson)

___
*[pkg/sql/conn_executor.go, line 1443 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MC7gwxE3JS7dmn7VQFS:-MDHgSUgEU2GrA9inMfQ:bzaq7im) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1adb0dd709b77e410997d597a3a993a5cab14/pkg/sql/conn_executor.go#L1443)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

comment is better now

You're sure that postgres allows you to `Execute` a portal multiple times after the portal is suspended, right?
</blockquote></details>

Yes, that's the case for statements of `tree.Rows` statement type.

I also updated the code to return errors similar to what Postgres does, and AFAICT this PR doesn't introduce any new deviations from PG.
___
*[pkg/sql/pgwire/testdata/pgtest/portals, line 464 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MBoix5oE7MSkm0By0ee:-MDHg67S5EBSO-7SPjV1:b1prvl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9a1adb0dd709b77e410997d597a3a993a5cab14/pkg/sql/pgwire/testdata/pgtest/portals#L464)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

If this question is for me, I think this is the time to figure out the exact details of the pg implementation. Obviously, the protocol is only defined by the Postgres code :0
I don't understand what the diff here says, but I think the one below seems bad
</blockquote></details>

I looked into the postgres code, and - in a sense - they store the number of "affected rows" in such cases so that when executing exhausted portals the corresponding number is used in the command tag. I updated the code to have such behavior in CRDB as well.
___
*[pkg/sql/pgwire/testdata/pgtest/portals, line 346 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MCw_zI6Bh7PSceyRise:-MDHGKEjCT6ytjOPVOff:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44563956f320d4e86424b6f37919a4545f62b24f/pkg/sql/pgwire/testdata/pgtest/portals#L346)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

instead of click-bait, please put some words about what we're testing
</blockquote></details>

Done.
___
*[pkg/sql/pgwire/testdata/pgtest/portals, line 396 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MCwaBWm2vQSDkobFv94:-MDHGv4EBCQRxOhO4InU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44563956f320d4e86424b6f37919a4545f62b24f/pkg/sql/pgwire/testdata/pgtest/portals#L396)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'd try another `Execute` here. I wanna make sure that pg returns `SELECT 0` multiple times.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Hey @andreimatei, @mjibson I wonder if you guys could take another look at this PR. I'm satisfied with the current way of things, so maybe it'll be the final review. I know I have taken my time working on this, but I'm hoping we could merge it sooner rather than later.

mjibson commented on pkg/sql/pgwire/command_result.go:
```diff
@@ -327,11 +328,16 @@ func (c *conn) newCommandResult(
 	}
 	telemetry.Inc(sqltelemetry.PortalWithLimitRequestCounter)
+	suspensionStrategy := nonSelectPortalSuspensionStrategy
+	if stmtTag == "SELECT" {
```
I don't think this is correct. See https://github.com/postgres/postgres/blob/2f48ede080f42b97b594fb14102c82ca1001b80c/src/backend/tcop/pquery.c#L227. There it looks at two things to determine the ONE_SELECT portal strategy: it must have a commandType of CMD_SELECT and not be have a modifying CTE (https://github.com/postgres/postgres/blob/2f48ede080f42b97b594fb14102c82ca1001b80c/src/backend/parser/parse_cte.c#L147). Here we are only checking for the tag, not the commandType.

mjibson commented on pkg/sql/pgwire/command_result.go:
```diff
@@ -345,6 +351,27 @@ func (c *conn) newMiscResult(pos sql.CmdPos, typ completionMsgType) *commandResu
+	// portal that consists of a SELECT statement. The execution of such
+	// statements can be "truly" suspended, so PG emits the actual number of
+	// returned rows since the previous suspension (if there was one).
+	selectPortalSuspensionStrategy portalSuspensionStrategy = iota
```
These comments should document how, or if, they relate to which enum values in postgres.

andreimatei reviewed (commented):
It's hard for me to review this without a description in the commit message of how postgres deals with various types of portals, and how we deal with them. I think you've added a lot of tests, which is great, but they're not readable enough for me to just use that.
I'd like to understand the behavior of `selects`, `update returning`, `insert` both with respect to what happens when they're resumed after being exhauster and with respect to the "affected rows" tags that they return after each suspension.
Btw, as I read in the pg source you've linked, it seems that for `update returning` pg executes the thing to completion buffering the returned values, and then it resuming the portal just consumes from the buffer, right? Whereas we actually suspend the execution of the statement, right? If so, results differ between us and pg when the txn is committed without exhausting such a portal, right (pg will update more rows than we do)?

I'm also happy not reviewing this if you prefer.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-:-MDqGWQU414uNvFVQBLV:bdmuq2c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @mjibson, and @yuzefovich)

___
*[pkg/sql/conn_executor.go, line 1477 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MDq8LuPBagfCjtG5LKx:-MDq8LuQAnU_94oxslp-:b-2wkwdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/092705ee7817aefcaac13ceb5996e71b93419116/pkg/sql/conn_executor.go#L1477)):*
> ```Go
> 			// In order to prevent such behavior, we check whether the portal
> 			// has been exhausted and execute the statement only if it hasn't.
> 			if portal.exhausted {
> ```

please move the exhausted portal code out of this function, to an `execExhaustedPortal()` or so
___
*[pkg/sql/conn_executor.go, line 1488 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MDq7mVL-QFRiBzemQVB:-MDq7mVL-QFRiBzemQVC:bqkbblx) ([raw file](https://github.com/cockroachdb/cockroach/blob/092705ee7817aefcaac13ceb5996e71b93419116/pkg/sql/conn_executor.go#L1488)):*
> ```Go
> 					// transitions which will emit CommandComplete messages and
> 					// alike (in a sense, by not calling execStmt we "execute"
> 					// the portal in such a way that it return 0 rows).
> ```

s/return/returns
___
*[pkg/sql/pgwire/command_result.go, line 478 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MDqFGUl1OOpcBkkfNmO:-MDqFGUl1OOpcBkkfNmP:b-ddcijd) ([raw file](https://github.com/cockroachdb/cockroach/blob/092705ee7817aefcaac13ceb5996e71b93419116/pkg/sql/pgwire/command_result.go#L478)):*
> ```Go
> 			switch r.suspensionStrategy {
> 			case selectPortalSuspensionStrategy:
> 				// In order to get the correct command tag for a SELECT
> ```

This comment doesn't help very much.
What would be great here is talking about which portals' results return the number of rows returned from the beginning, and which portals return the number of rows part of the current `ExecPortal`. Like, am I right to understand that `select * from foo` executed with limit 100 will return a tag saying 100 rows with every resume, but `update foo returning bar` executed with limit 100 will return tags 100, then 200, then 300?


<!-- Sent from Reviewable.io -->

mjibson commented:
I think we're ok even in the `update returning` case because we only allow a single portal to be opened at any time, which means a user could not partially execute a cockroach update, execute something else, and then return to the update. It is also possible the optimizer does some stuff that executes all mutations first so it is correct anyway? Not sure.

yuzefovich committed:
sql: fix portals after exhausting rows

Previously, we would erroneously restart the execution from the very
beginning of empty, unclosed portals after they have been fully
exhausted when we should be returning no rows or an error in such
scenarios. This is now fixed by tracking whether a portal is exhausted
or not and intercepting the calls to `execStmt` when the conn executor
state machine is in an open state.

Note that the current solution has known deviations from Postgres:
- when attempting to execute portals of statements that don't return row
sets, on the second and consequent attempt PG returns an error while we
are silently doing nothing (meaning we do not run the statement at all
and return 0 rows)
- we incorrectly populate "command tag" field of pgwire messages of some
rows-returning statements after the portal suspension (for example,
a suspended UPDATE RETURNING in PG will return the total count of "rows
affected" while we will return the count since the last suspension).

These deviations are deemed acceptable since this commit fixes a much
worse problem - re-executing an exhausted portal (which could be
a mutation meaning, previously, we could have executed a mutation
multiple times).

The reasons for why this commit does not address these deviations are:
- Postgres has a concept of "portal strategy"
(see https://github.com/postgres/postgres/blob/2f9661311b83dc481fc19f6e3bda015392010a40/src/include/utils/portal.h#L89).
- Postgres has a concept of "command" type (these are things like
SELECTs, UPDATEs, INSERTs, etc,
see https://github.com/postgres/postgres/blob/1aac32df89eb19949050f6f27c268122833ad036/src/include/nodes/nodes.h#L672).

CRDB doesn't have these concepts, and without them any attempt to
simulate Postgres results in a very error-prone and brittle code.

Release note (bug fix): Previously, CockroachDB would erroneously
restart the execution of empty, unclosed portals after they have been
fully exhausted, and this has been fixed.

yuzefovich reviewed (commented):
I've talked with Matt, and we decided that the new additions that attempt to simulate the "rows affected" behavior in the command tag are incorrect and very brittle, so that code has been deleted. The current PR fixes the really bad problem (of re-executing exhausted portals) but introduces some deviations which are less severe. We found such trade-off acceptable.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-:-MDq_KOn7qknXdJzXSNi:by4x2gf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @mjibson)

___
*[pkg/sql/conn_executor.go, line 1477 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MDq8LuPBagfCjtG5LKx:-MDqQAw53K3tACjdFOAq:b-hoy7fm) ([raw file](https://github.com/cockroachdb/cockroach/blob/092705ee7817aefcaac13ceb5996e71b93419116/pkg/sql/conn_executor.go#L1477)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please move the exhausted portal code out of this function, to an `execExhaustedPortal()` or so
</blockquote></details>

The code here has been simplified, but I also extracted a separate function to not flood this file with long comments.
___
*[pkg/sql/conn_executor.go, line 1488 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MDq7mVL-QFRiBzemQVB:-MDqQ8bs502wk7TcXAjO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/092705ee7817aefcaac13ceb5996e71b93419116/pkg/sql/conn_executor.go#L1488)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/return/returns
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor.go, line 1489 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7FRxi84OqPVll_sUME-r6-1489:-MDqDIfb7Xdf2nAW6z_3:b-yofr9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/092705ee7817aefcaac13ceb5996e71b93419116/pkg/sql/conn_executor.go#L1489)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

What about statements like `VALUES (1)`?
</blockquote></details>

It's an interesting point. I was a little bit surprised, but this PR behaves exactly like Postgres because a query like `VALUES (0)` is actually parsed as a `tree.Select` object in which `Select` parameter is a `tree.ValuesClause`. Added it as a test.

I'm assuming that PG's `CMD_SELECT` is equivalent to ours `Stmt.AST.StatementTag() == "SELECT" (which seems like a reasonable assumption given https://github.com/postgres/postgres/blob/1aac32df89eb19949050f6f27c268122833ad036/src/include/nodes/nodes.h#L67

Update: this code has been removed.
___
*[pkg/sql/conn_executor.go, line 1510 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-M7FRxi84OqPVll_sUME-r6-1510:-MDqG4MV1mq3x_7gP025:b-clh693) ([raw file](https://github.com/cockroachdb/cockroach/blob/092705ee7817aefcaac13ceb5996e71b93419116/pkg/sql/conn_executor.go#L1510)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

This comment is slightly misleading. It should describe how, while we do support portal suspension, it happens in another state machine and thus once execStmt has returned, the portal has executed to completion.
</blockquote></details>

Good point, updated.
___
*[pkg/sql/pgwire/command_result.go, line 332 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MDHg685tv-o2FCapZoE-r6-332:-MDqIB56Ex9uIowwqN10:b-pmrpk4) ([raw file](https://github.com/cockroachdb/cockroach/blob/092705ee7817aefcaac13ceb5996e71b93419116/pkg/sql/pgwire/command_result.go#L332)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I don't think this is correct. See https://github.com/postgres/postgres/blob/2f48ede080f42b97b594fb14102c82ca1001b80c/src/backend/tcop/pquery.c#L227. There it looks at two things to determine the ONE_SELECT portal strategy: it must have a commandType of CMD_SELECT and not be have a modifying CTE (https://github.com/postgres/postgres/blob/2f48ede080f42b97b594fb14102c82ca1001b80c/src/backend/parser/parse_cte.c#L147). Here we are only checking for the tag, not the commandType.
</blockquote></details>

I agree, the current code deviates from Postgres.

However, understanding whether a query has a modifying CTE (or a subquery) seems quite hard at the point when `newCommandResult` is called - it is done before `makeOptimizerPlan`, and I feel like a minor deviation like this should not block this PR. Namely, the deviation is minor because this `suspensionStrategy` currently only influences "rows affected" parameter which is need to populate correctly "command tag".

Update: this code has been removed.
___
*[pkg/sql/pgwire/command_result.go, line 366 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MDHg685tv-o2FCapZoE-r6-366:-MDqSwt7-yZyUx2KvVlR:b69kmyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/092705ee7817aefcaac13ceb5996e71b93419116/pkg/sql/pgwire/command_result.go#L366)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

These comments should document how, or if, they relate to which enum values in postgres.
</blockquote></details>

This code has been removed.
___
*[pkg/sql/pgwire/command_result.go, line 478 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MDqFGUl1OOpcBkkfNmO:-MDqSziLE5tyAkmOJSfg:b69kmyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/092705ee7817aefcaac13ceb5996e71b93419116/pkg/sql/pgwire/command_result.go#L478)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

This comment doesn't help very much.
What would be great here is talking about which portals' results return the number of rows returned from the beginning, and which portals return the number of rows part of the current `ExecPortal`. Like, am I right to understand that `select * from foo` executed with limit 100 will return a tag saying 100 rows with every resume, but `update foo returning bar` executed with limit 100 will return tags 100, then 200, then 300?
</blockquote></details>

This code has been removed.


<!-- Sent from Reviewable.io -->

mjibson commented on pkg/sql/conn_executor_prepare.go:
```diff
@@ -397,6 +394,17 @@ func (ex *connExecutor) addPortal(
 }
 
+// exhaustPortal marks a portal with the provided name as "exhausted" and
+// returns an error if there is no portal with such name.
```
returns an error -> panics

mjibson commented:
Very +1 with allowing slightly differing behavior so that we can fix the bad reexec bug.

yuzefovich reviewed (commented):
I'll wait for @andreimatei to take another quick look. Let me know if you think it's not necessary.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-:-MDqpoXkDpSnwRV5zCRr:be7nssg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @mjibson)

___
*[pkg/sql/conn_executor_prepare.go, line 398 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/48842#-MCiQpzEmIWnnZpmoSGr-r7-398:-MDqpbc200atA5QUWXvB:b1xlfn7) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc791f8196dbb24d06d4638db4d304b31c0dc8e9/pkg/sql/conn_executor_prepare.go#L398)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

returns an error -> panics
</blockquote></details>

Oops, indeed. I have Goland collapsing such statements, and usually those are "return error".


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Hey @andreimatei, are you intending to take a look? That's ok if not - I think this PR has received a fair amount of scrutiny, and I'd like to merge it tomorrow.

andreimatei commented:
Go ahead

On Wed, Aug 5, 2020, 6:53 PM yuzefovich <notifications@github.com> wrote:

> Hey @andreimatei <https://github.com/andreimatei>, are you intending to
> take a look? That's ok if not - I think this PR has received a fair amount
> of scrutiny, and I'd like to merge it tomorrow.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/48842#issuecomment-669585467>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC4C4N7CMYL2REWFJBFRPTR7HPF3ANCNFSM4NAGUTJQ>
> .
>

yuzefovich commented:
Thanks everyone for the input!

bors r+
